### PR TITLE
#2101 inherited properties need to be analysed against redefined properties when inheriting mappings

### PR DESCRIPTION
**Pull request by sjaakd**
Possibly need to add some unit test.. Although quite some testcases failed when I implemented it wrong 😄

### CODE REVIEW

**Comment by sjaakd** [comment]
@filiphr .. shall we add the reproducer anyway..? Always good to have additional UT.

**Comment by filiphr** [comment]
This is a tricky one. Basically what happens it the following mapper:

```
@Mapper
public interface Issue2101Mapper {
    @Mapping(target = "value1", source = "codeValue1")
    @Mapping(target = "value2", source = "codeValue2")
    Source map(Target target);

    //@InheritInverseConfiguration
    @Mapping(target = "codeValue1", source = "value1")
    @Mapping(target = "codeValue2", source = "value2")
    @Mapping(target = "codeValue1.code", constant = "c1")
    @Mapping(target = "codeValue1.value", source = "value1")
    @Mapping(target = "codeValue2.code", constant = "c2")
    @Mapping(target = "codeValue2.value", source = "value2")
    Target map(Source source);
}
```

Perhaps when doing the inverse inheritance if there is a target mapping which has the reversed segment we should not use it.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java [hunk 1]
```diff
@@ -177,13 +177,54 @@ public void applyInheritedOptions(SourceMethod templateMethod, boolean isInverse
...
-            mappings.addAll( newMappings );
+            addAllNonRedefined( newMappings );
 
             // filter new mappings
             filterNestedTargetIgnores( mappings );
         }
     }
 
+    private void addAllNonRedefined(Set<MappingOptions> inheritedMappings) {
+        Set<String> redefinedSources = mappings.stream()
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java:188]: Streams are all OK, but we are going twice through the same mappings.

Why not defined the sets, then use a for-loop to iterate the mappings and add to the sets?

Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java:188]: will correct.. I'm all into this streams nowadays.. 😄

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java:188]: I love streams as well. However, good old imperative style is better sometimes. Lately I am all into performance 😀

**Pull request merged**